### PR TITLE
Add support to create Application Engine(UCP) context

### LINKED ISSUE #488: Add support to create UCP context
**Describe the feature request**
CLI should support the new Context type "Application Engine"(Unified Control Plane). The context created should have reference data for the CSP authentication and to start with should also create a kubeconfig file pointing to "org" resource in UCP resource hierarchy(Org/project/space) and once the kubeconfig is created , the kubeconfig path and kubeconfig context references should also be store in the Context created.

**Describe alternative(s) you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by prkalle**
### What this PR does / why we need it
This PR adds support to create Application Engine(UCP) context
Changes Summary:
- Add support to create Application Engine(UCP) context
- Removed the implemetation for mission-control selfmanaged context creation
- UX changes for prompt based context creation

TODO: 
- Modification of kubeconfig generation to incorporate ExecConfig instead of using the access-token would be done in the follow up PR.  
- Update go.mod and go.sum after the tanzu-plugin-runtime PR is merged
### Which issue(s) this PR fixes

Fixes #488 

### Describe testing done for PR

Created the UCP context by specifying the context "type" as "application-engine" and also by providing the endpoint
```
❯ ./bin/tanzu context create myucp --endpoint https://api-dev.tanzu.cloud.vmware.com --type application-engine  --staging
[i] API token env var is set

[ok] successfully created a Application Engine(UCP) context


#context list shows the context created as below

❯ ./bin/tanzu context list
Target:  kubernetes
  NAME                       ISACTIVE  ENDPOINT                                                                         KUBECONFIGPATH                           KUBECONTEXT
  myucp                      true      https://api-dev.tanzu.cloud.vmware.com/org/bc27608b-4809-4cac-9e04-778803963da2  /Users/pkalle/.kube/config               tanzu-cli-myucp

// Kubeconfig after UCP context is created is shown below
apiVersion: v1
clusters:
- cluster:
    server: https://api-dev.tanzu.cloud.vmware.com/org/bc27608b-4809-4cac-9e04-778803963da2
  name: tanzu-cli-myucp/current
contexts:
- context:
    cluster: tanzu-cli-myucp/current
    user: tanzu-cli-myucp-user
  name: tanzu-cli-myucp
current-context: tanzu-cli-myucp
kind: Config
preferences: {}
users:
- name: tanzu-cli-myucp-user
  user:
    token: <REDACTED>
```
Tried to create the same context again and CLI returned the error as expected

```
❯ ./bin/tanzu context create myucp --endpoint https://api-dev.tanzu.cloud.vmware.com --type application-engine  --staging
[x] : context "myucp" already exists
```

Create the UCP context by providing the `--endpoint`, `--endpoint-ca-certificate` successfully

```
❯ ./bin/tanzu context delete myucp -y
[i] Deleting entry for cluster myucp
❯ ./bin/tanzu context create myucp --endpoint https://api-dev.tanzu.cloud.vmware.com --type application-engine  --staging --endpoint-ca-certificate ~/temp/cluster_ca_cert.crt
[i] API token env var is set

[ok] successfully created a Application Engine(UCP) context


// Kubeconfig after UCP context is created is shown below
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <REDACTED>
    server: https://api-dev.tanzu.cloud.vmware.com/org/bc27608b-4809-4cac-9e04-778803963da2
  name: tanzu-cli-myucp/current
contexts:
- context:
    cluster: tanzu-cli-myucp/current
    user: tanzu-cli-myucp-user
  name: tanzu-cli-myucp
current-context: tanzu-cli-myucp
kind: Config
preferences: {}
users:
- name: tanzu-cli-myucp-user
  user:
    token: <REDACTED>

```

Create the UCP context by providing the `--endpoint`, `--insecure-skip-tls-verify` successfully. The kubeconfig cluster's structure would be updated with `insecure-skip-tls-verify` so that kubectl would ignore the endpoint cert verification while interacting with UCP endpoint

```
❯ ./bin/tanzu context create myucp --endpoint https://api-dev.tanzu.cloud.vmware.com --type application-engine  --staging --insecure-skip-tls-verify
[i] API token env var is set

[ok] successfully created a Application Engine(UCP) context
[i] Checking for required plugins...

// Kubeconfig after UCP context is created is shown below
apiVersion: v1
clusters:
- cluster:
    insecure-skip-tls-verify: true
    server: https://api-dev.tanzu.cloud.vmware.com/org/bc27608b-4809-4cac-9e04-778803963da2
  name: tanzu-cli-myucp/current
contexts:
- context:
    cluster: tanzu-cli-myucp/current
    user: tanzu-cli-myucp-user
  name: tanzu-cli-myucp
current-context: tanzu-cli-myucp
kind: Config
preferences: {}
users:
- name: tanzu-cli-myucp-user
  user:
    token:  <REDACTED>
```

UCP context creation using prompts
```
// Step 1: (select Application Engine from the list)
❯ ./bin/tanzu context create --staging
? Select context creation type  [Use arrows to move, type to filter]
  Mission Control
> Application Engine
  Kubernetes Cluster Endpoint
  Kubernetes Local Kubeconfig

//Step 2:  (Enter the control plane endpoint, specify the value if endpoint is not same as default value provided)
❯ ./bin/tanzu context create --staging
? Select context creation type Application Engine
? Enter control plane endpoint (https://api.tanzu.cloud.vmware.com) https://api-dev.tanzu.cloud.vmware.com

// Step 3: (Provide the context name)
❯ ./bin/tanzu context create --staging
? Select context creation type Application Engine
? Enter control plane endpoint https://api-dev.tanzu.cloud.vmware.com
? Give the context a name myucp
[i] API token env var is set

[ok] successfully created a Application Engine(UCP) context
[i] Checking for required plugins...

// If user wants to provide the endpoint CA certificagte or skipTLSVerify, User should provide them as command line options as shown below
❯ ./bin/tanzu context create --staging --endpoint-ca-certificate ~/temp/cluster_ca_cert.crt
? Select context creation type  [Use arrows to move, type to filter]
> Mission Control
  Application Engine
  Kubernetes Cluster Endpoint
  Kubernetes Local Kubeconfig
```

### Release note

```release-note
Add support to create Application Engine (UCP) context
```

Create a UCP context using prompt

#### Special notes for your reviewer

This PR has dependency on tanzu-plugin-runtime [PR97](https://github.com/vmware-tanzu/tanzu-plugin-runtime/pull/97)

### CODE REVIEW

File: pkg/command/context.go [hunk 1]
```diff
@@ -129,6 +136,7 @@ func initCreateCtxCmd() {
 	createCtxCmd.Flags().BoolVar(&staging, "staging", false, "use CSP staging issuer")
 	createCtxCmd.Flags().StringVar(&endpointCACertPath, "endpoint-ca-certificate", "", "path to the endpoint public certificate")
 	createCtxCmd.Flags().BoolVar(&skipTLSVerify, "insecure-skip-tls-verify", false, "skip endpoint's TLS certificate verification")
+	createCtxCmd.Flags().StringVar(&contextType, "type", "", "type of context to create (mission-control | application-engine | k8s-cluster-endpoint | k8s-local-kubeconfig)")
```
Comment (prkalle) [inline pkg/command/context.go:139]: Would request comments/suggestion on the context type names

File: pkg/command/context.go [hunk 2]
```diff
@@ -212,33 +223,41 @@ func createNewContext() (context *configtypes.Context, err error) {
 		return createContextWithKubeconfig()
 	}
 	// user provided command line options to create a context using endpoint
-	if endpoint != "" || selfManaged {
-		return createContextWithEndpoint()
+	if contextType == "application-engine" || (endpoint != "" && isGlobalUCPEndpoint(endpoint)) {
```
Comment (vuil) [inline pkg/command/context.go:226]: nit: suggesting a bit of refacting: how about defining the various creation type strings (""Mission Control", "Kubernetes Local Kubeconfig".. referenced in L250+) as constants, then also use them here to possibly assign to a local var initialized to "", call promptContextType only if we cannot detect what to use, and in all cases just return the result of createContextUsingContextType?

File: pkg/auth/ucp/kubeconfig.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	"strings"
+
+	"github.com/pkg/errors"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	kubeutils "github.com/vmware-tanzu/tanzu-cli/pkg/auth/utils/kubeconfig"
+	configtypes "github.com/vmware-tanzu/tanzu-plugin-runtime/config/types"
+)
+
+func GetUCPKubeconfig(c *configtypes.Context, endpoint, orgID, endpointCACertPath string, skipTLSVerify bool) (string, string, string, error) {
```
Comment (vuil) [inline pkg/auth/ucp/kubeconfig.go:20]: nit: a function header would be useful here

File: pkg/auth/ucp/kubeconfig.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	clusterCACertData := ""
+	if endpointCACertPath != "" {
+		fileBytes, err := os.ReadFile(endpointCACertPath)
+		if err != nil {
+			return "", "", "", errors.Wrapf(err, "error reading CA certificate file %s", endpointCACertPath)
+		}
+		clusterCACertData = base64.StdEncoding.EncodeToString(fileBytes)
+	}
+
+	contextName := "tanzu-cli-" + c.Name
```
Comment (vuil) [inline pkg/auth/ucp/kubeconfig.go:36]: Nit: might be good to write some tiny helper functions to construct these, and use them where appropriate.

File: pkg/auth/ucp/kubeconfig_test.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+				user := config.AuthInfos[config.Contexts[kubeContext].AuthInfo]
+
+				Expect(cluster.Server).To(Equal(clusterAPIServerURL))
+				Expect(config.Contexts[kubeContext].AuthInfo).To(Equal("tanzu-cli-" + ucpContext.Name + "-user"))
+				Expect(gotClusterName).To(Equal("tanzu-cli-" + ucpContext.Name + "/current"))
+				Expect(len(cluster.CertificateAuthorityData)).ToNot(Equal(0))
+				Expect(user.Token).To(Equal(ucpContext.GlobalOpts.Auth.AccessToken))
+			})
+		})
+		Context("When skipTLSVerify is true", func() {
```
Comment (vuil) [inline pkg/auth/ucp/kubeconfig_test.go:98]: nit: the only difference between this and previous test is that (+) no cert path is provided and certauthdata is set to empty as a result. Given current implementation the behavior of both tests are the same regardless of whether skipTLSVerify is true or not. So maybe this Context and It string should be updated to more accurately reflect (+)

File: pkg/command/context.go [hunk 6]
```diff
@@ -101,14 +99,23 @@ var createCtxCmd = &cobra.Command{
...
 
     # Create a TKG management cluster context by explicit request to skip TLS verification, which is insecure:
-    tanzu context create --endpoint https://k8s.example.com[:port] --insecure-skip-tls-verify
+    tanzu context create mgmt-cluster --endpoint https://k8s.example.com[:port] --insecure-skip-tls-verify
 
     # Create a TKG management cluster context using default kubeconfig path and a kubeconfig context
     tanzu context create mgmt-cluster --kubecontext kubecontext
 
+    # Create a Application Engine(UCP) context
+    tanzu context create myucp -endpoint https://k8s.example.com[:port] --type application-engine
```
Comment (vuil) [inline pkg/command/context.go:111]: should the URL be updated?

File: pkg/command/context.go [hunk 7]
```diff
@@ -101,14 +99,23 @@ var createCtxCmd = &cobra.Command{
     tanzu context create mgmt-cluster --kubeconfig path/to/kubeconfig --kubecontext kubecontext
 
     # Create a TKG management cluster context by using the provided CA Bundle for TLS verification:
```
Comment (vuil) [inline pkg/command/context.go:101]: I feel there should be some mention of how the CLI would attempt to detect the context type where possible. 
Coupled with that we either show what you have below and mentioned that `--type` might be omittable, OR show examples without the `--type` then mention that `--type` might need to be added in some cases.

File: pkg/command/context.go [hunk 8]
```diff
@@ -419,32 +448,22 @@ func globalLogin(c *configtypes.Context) (err error) {
...
-	issuer, err := getIssuerURLForTMCEndPoint(c.GlobalOpts.Endpoint)
+func globalUCPLogin(c *configtypes.Context) error {
+	claims, err := doCSPAuthAndUpdateContext(c, "UCP")
 	if err != nil {
 		return err
 	}
-	refreshToken := ""
-	token, err := csp.GetAccessTokenFromSelfManagedIDP(refreshToken, issuer)
+	//TODO(prkalle) to update "ucpOrgID" to use plugin runtime constant
+	c.AdditionalMetadata["ucpOrgID"] = claims.OrgID
```
[the remark this answers is not in the document]
Comment (prkalle) [reply pkg/command/context.go:457]: No, I am considering `ucpOrdID` `ucpProjectName` and `ucpSpaceName` as metadata for UCP context, hence didn't add OrgID field to Context.

File: pkg/command/context.go (hunk 2 above)
Comment (prkalle) [reply pkg/command/context.go:226]: Thanks! updated.

File: pkg/auth/ucp/kubeconfig_test.go (hunk 5 above)
Comment (prkalle) [reply pkg/auth/ucp/kubeconfig_test.go:98]: Updated the Context description of the both the tests to more accurately reflect the test behavior.

File: pkg/command/context.go (hunk 6 above)
Comment (prkalle) [reply pkg/command/context.go:111]: updated the URL

File: pkg/command/context.go (hunk 7 above)
Comment (prkalle) [reply pkg/command/context.go:101]: Thanks! Updated the examples with and without `--type` and mentioned that `--type` is optional and if not provided would be inferred from the endpoint in some cases.

File: pkg/command/context.go [hunk 9]
```diff
@@ -94,21 +100,36 @@ var createCtxCmd = &cobra.Command{
...
 
     # Create a TKG management cluster context by explicit request to skip TLS verification, which is insecure:
-    tanzu context create --endpoint https://k8s.example.com[:port] --insecure-skip-tls-verify
+    tanzu context create mgmt-cluster --endpoint https://k8s.example.com[:port] --insecure-skip-tls-verify
 
     # Create a TKG management cluster context using default kubeconfig path and a kubeconfig context
     tanzu context create mgmt-cluster --kubecontext kubecontext
 
+   # Create a Application Engine(UCP) context with default endpoint (--type option is not necessary for default endpoint)
+    tanzu context create myucp -endpoint https://api.tanzu.cloud.vmware.com
```
Comment (vuil) [inline pkg/command/context.go:122]: nit: trailing space

File: pkg/command/context.go (hunk 1 above)
Comment (vuil) [reply pkg/command/context.go:153]: let's go with these; we can update in a followup if necessary.

File: pkg/command/context.go [hunk 10]
```diff
@@ -94,21 +100,36 @@ var createCtxCmd = &cobra.Command{
...
+
+    # Create a Application Engine(UCP) context (--type option is needed for non-default endpoint)
+    tanzu context create myucp -endpoint https://non-default.ucp.endpoint.com --type application-engine
+
+    # Create a Application Engine(UCP) context by using the provided CA Bundle for TLS verification:
+    tanzu context create myucp --endpoint https://api.tanzu.cloud.vmware.com  --endpoint-ca-certificate /path/to/ca/ca-cert
+
+    # Create a Application Engine(UCP) context by explicit request to skip TLS verification, which is insecure:
+    tanzu context create myucp --endpoint https://api.tanzu.cloud.vmware.com --insecure-skip-tls-verify
+
```
Comment (vuil) [inline pkg/command/context.go:132]: preexisting: we should add example for mission-control as well, but I expect we will have to do another pass on the examples, we can do this in the followup.

**Review by vuil — APPROVED** [review]
lgtm, thanks for the changes!

File: pkg/command/context.go (hunk 9 above)
Comment (prkalle) [reply pkg/command/context.go:122]: updated

**Pull request merged**